tests/components/slack/test_notify.py

```diff
@@ -1,40 +1,63 @@
 """Test slack notifications."""
+import copy
 import logging
+from typing import List
 from unittest.mock import AsyncMock, Mock, patch
 
 from _pytest.logging import LogCaptureFixture
 import aiohttp
 from slack.errors import SlackApiError
 
+from homeassistant.components import notify
+from homeassistant.components.slack import DOMAIN
 from homeassistant.components.slack.notify import (
     CONF_DEFAULT_CHANNEL,
     SlackNotificationService,
-    async_get_service,
 )
-from homeassistant.const import CONF_API_KEY, CONF_ICON, CONF_USERNAME
+from homeassistant.const import CONF_API_KEY, CONF_ICON, CONF_PLATFORM, CONF_USERNAME
 from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.setup import async_setup_component
 
 MODULE_PATH = "homeassistant.components.slack.notify"
 
-
-async def test_get_service(hass: HomeAssistantType, caplog: LogCaptureFixture):
-    """Test async_get_service with exceptions."""
-    config = {
-        CONF_API_KEY: "12345",
-        CONF_DEFAULT_CHANNEL: "channel",
-    }
-
-    with patch(MODULE_PATH + ".aiohttp_client") as mock_session, patch(
-        MODULE_PATH + ".WebClient"
+DEFAULT_CONFIG = {
+    notify.DOMAIN: [
+        {
+            CONF_PLATFORM: DOMAIN,
+            CONF_API_KEY: "12345",
+            CONF_DEFAULT_CHANNEL: "channel",
+        }
+    ]
+}
+
+
+def filter_log_records(caplog: LogCaptureFixture) -> List[logging.LogRecord]:
+    """Filter all unrelated log records."""
+    return [
+        rec for rec in caplog.records if rec.name.endswith(f"{DOMAIN}.{notify.DOMAIN}")
+    ]
+
+
+async def test_setup(hass: HomeAssistantType, caplog: LogCaptureFixture):
+    """Test setup slack notify."""
+    config = DEFAULT_CONFIG
+
+    with patch(
+        MODULE_PATH + ".aiohttp_client",
+        **{"async_get_clientsession.return_value": (session := Mock())},
+    ), patch(
+        MODULE_PATH + ".WebClient",
+        return_value=(client := AsyncMock()),
     ) as mock_client, patch(
-        MODULE_PATH + ".SlackNotificationService"
+        MODULE_PATH + ".SlackNotificationService",
+        return_value=(service := AsyncMock()),
     ) as mock_slack_service:
-        mock_session.async_get_clientsession.return_value = session = Mock()
-        mock_client.return_value = client = AsyncMock()
 
-        # Normal setup
-        mock_slack_service.return_value = service = Mock()
-        assert await async_get_service(hass, config) == service
+        await async_setup_component(hass, notify.DOMAIN, config)
+        await hass.async_block_till_done()
+        assert hass.data[notify.NOTIFY_SERVICES][DOMAIN][0] == service
```

Comment:
We don't want to assert any integration details, like `hass.data`. When the notify service is registered in the notify integration it will register a service in the core service registry. So we can use the service registry to check if a service is registered. Use eg `hass.services.has_service`.

Reply:
I removed the calls to `hass.data`. My primary goal with these test was to assert that the `SlackNotificationService` class itself is called with the correct arguments. For that reason I replaced it with a mocked version -> `mock_slack_service`. However, doing so also means that the service is never truly registered.

Using `hass.data` was a small workaround to that which isn't really necessary.

Reply:
Yeah. We shouldn't patch `SlackNotificationService` but let it register.